File: service/src/main/java/no/nav/melosys/service/behandling/VilkaarsresultatService.java
package no.nav.melosys.service.behandling;

import java.util.*;
import java.util.stream.Collectors;

import no.nav.melosys.domain.Behandlingsresultat;
import no.nav.melosys.domain.VilkaarBegrunnelse;
import no.nav.melosys.domain.Vilkaarsresultat;
import no.nav.melosys.domain.kodeverk.Kodeverk;
import no.nav.melosys.domain.kodeverk.Vilkaar;
import no.nav.melosys.exception.FunksjonellException;
import no.nav.melosys.exception.IkkeFunnetException;
import no.nav.melosys.repository.BehandlingsresultatRepository;
import no.nav.melosys.service.saksbehandling.SaksbehandlingRegler;
import no.nav.melosys.service.vilkaar.VilkaarDto;
import org.springframework.stereotype.Service;
import org.springframework.transaction.annotation.Transactional;

import static no.nav.melosys.domain.kodeverk.Vilkaar.*;
import static no.nav.melosys.service.behandling.BehandlingsresultatService.KAN_IKKE_FINNE_BEHANDLINGSRESULTAT;

@Service
public class VilkaarsresultatService {

    private final BehandlingsresultatRepository behandlingsresultatRepository;
    private final SaksbehandlingRegler saksbehandlingRegler;

    public static final Collection<Vilkaar> IMMUTABLE_VILKAAR = Collections.singleton(FO_883_2004_INNGANGSVILKAAR);

    public VilkaarsresultatService(
        BehandlingsresultatRepository behandlingsresultatRepository,
        SaksbehandlingRegler saksbehandlingRegler) {
        this.behandlingsresultatRepository = behandlingsresultatRepository;
        this.saksbehandlingRegler = saksbehandlingRegler;
    }

    public Behandlingsresultat hentBehandlingsresultat(long behandlingsid) {
        return behandlingsresultatRepository.findById(behandlingsid)
            .orElseThrow(() -> new IkkeFunnetException(KAN_IKKE_FINNE_BEHANDLINGSRESULTAT + behandlingsid));
    }

    @Transactional(readOnly = true)
    public List<VilkaarDto> hentVilkaar(long behandlingID) {
        Behandlingsresultat behandlingsresultat = hentBehandlingsresultat(behandlingID);

        List<VilkaarDto> vilkaarDtoListe = new ArrayList<>();
        for (Vilkaarsresultat vilkaarsresultat : behandlingsresultat.getVilkaarsresultater()) {
            VilkaarDto vilkaarDto = new VilkaarDto();
            vilkaarDto.setVilkaar(vilkaarsresultat.getVilkaar().getKode());
            vilkaarDto.setOppfylt(vilkaarsresultat.isOppfylt());
            vilkaarDto.setBegrunnelseKoder(vilkaarsresultat.getBegrunnelser().stream().map(VilkaarBegrunnelse::getKode).collect(Collectors.toSet()));
            vilkaarDto.setBegrunnelseFritekst(vilkaarsresultat.getBegrunnelseFritekst());
            vilkaarDto.setBegrunnelseFritekstEngelsk(vilkaarsresultat.getBegrunnelseFritekstEessi());
            vilkaarDtoListe.add(vilkaarDto);
        }

        return vilkaarDtoListe;
    }

    @Transactional(readOnly = true)
    public Optional<Vilkaarsresultat> finnVilkaarsresultat(long behandlingID, Vilkaar vilkaar) {
        return hentBehandlingsresultat(behandlingID).getVilkaarsresultater().stream()
            .filter(vilkaarsresultat -> vilkaarsresultat.getVilkaar().equals(vilkaar))
            .findFirst();
    }

    @Transactional(readOnly = true)
    public boolean oppfyllerVilkaar(long behandlingID, Vilkaar vilkaar) {
        return hentBehandlingsresultat(behandlingID).getVilkaarsresultater().stream()
            .anyMatch(vilkaarsresultat -> vilkaar.equals(vilkaarsresultat.getVilkaar()) && vilkaarsresultat.isOppfylt());
    }

    @Transactional(readOnly = true)
    public boolean harVilkaarForArtikkel12(long behandlingID) {
        Optional<Vilkaarsresultat> art121Vilkaar = finnVilkaarsresultat(behandlingID, FO_883_2004_ART12_1);
        Optional<Vilkaarsresultat> art122Vilkaar = finnVilkaarsresultat(behandlingID, FO_883_2004_ART12_2);
        return art121Vilkaar.isPresent() || art122Vilkaar.isPresent();
    }

    @Transactional(readOnly = true)
    public boolean harVilkaarForArtikkel16(long behandlingID) {
        Optional<Vilkaarsresultat> art161Vilkaar = finnVilkaarsresultat(behandlingID, FO_883_2004_ART16_1);
        return art161Vilkaar.isPresent();
    }

    @Transactional
    public void registrerVilkår(long behandlingID, List<VilkaarDto> vilkaarDtoer) {
        validerVilkår(vilkaarDtoer);
        Behandlingsresultat behandlingsresultat = hentBehandlingsresultat(behandlingID);
        tømVilkårsresultatFraBehandlingsresultat(behandlingID);
        // Flush fordi vi potensielt legger til samme vilkåret igjen. INSERT kommer før DELETE i Hibernate, som skaper UNIQUE constraint problemer uten flush.
        behandlingsresultatRepository.saveAndFlush(behandlingsresultat);

        for (VilkaarDto vilkaarDto : vilkaarDtoer) {
            Vilkaarsresultat vilkaarsresultat = lagVilkaarsresultat(
                behandlingsresultat,
                Vilkaar.valueOf(vilkaarDto.getVilkaar()),
                vilkaarDto.isOppfylt(),
                vilkaarDto.getBegrunnelseKoder(),
                vilkaarDto.getBegrunnelseFritekst(),
                vilkaarDto.getBegrunnelseFritekstEngelsk());
            behandlingsresultat.getVilkaarsresultater().add(vilkaarsresultat);
        }
        behandlingsresultatRepository.save(behandlingsresultat);
    }

    @Transactional
    public void tømVilkårsresultatFraBehandlingsresultat(long behandlingID) {
        Behandlingsresultat behandlingsresultat = hentBehandlingsresultat(behandlingID);

        var behandling = behandlingsresultat.getBehandling();
        var fagsak = behandling.getFagsak();
        if (fagsak.erSakstypeEøs() && !saksbehandlingRegler.harIngenFlyt(behandling)) {
            behandlingsresultat.getVilkaarsresultater().removeIf(vilkaarsresultat -> !IMMUTABLE_VILKAAR.contains(vilkaarsresultat.getVilkaar()));
        } else {
            behandlingsresultat.getVilkaarsresultater().clear();
        }
        behandlingsresultatRepository.saveAndFlush(behandlingsresultat);
    }

    private void validerVilkår(List<VilkaarDto> vilkaarDtoer) {

        final Collection<String> nyeVilkår = vilkaarDtoer.stream().map(VilkaarDto::getVilkaar).toList();

        for (Vilkaar immutableVilkår : IMMUTABLE_VILKAAR) {
            if (nyeVilkår.contains(immutableVilkår.getKode())) {
                throw new FunksjonellException("Kan ikke endre vilkår " + immutableVilkår);
            }
        }
    }

    @Transactional
    public void oppdaterVilkaarsresultat(long behandlingID,
                                         Vilkaar vilkaar,
                                         boolean oppfylt,
                                         Set<Kodeverk> begrunnelseKoder) {
        Behandlingsresultat behandlingsresultat = hentBehandlingsresultat(behandlingID);
        behandlingsresultat.getVilkaarsresultater().clear();
        // Flush fordi vi potensielt legger til samme vilkåret igjen. INSERT kommer før DELETE i Hibernate, som skaper UNIQUE constraint problemer uten flush.
        behandlingsresultatRepository.saveAndFlush(behandlingsresultat);
        Vilkaarsresultat vilkaarsresultat = lagVilkaarsresultat(
            behandlingsresultat,
            vilkaar,
            oppfylt,
            begrunnelseKoder.stream().map(Kodeverk::getKode).collect(Collectors.toSet()));

        behandlingsresultat.getVilkaarsresultater().add(vilkaarsresultat);
        behandlingsresultatRepository.save(behandlingsresultat);
    }

    private Vilkaarsresultat lagVilkaarsresultat(Behandlingsresultat behandlingsresultat,
                                                 Vilkaar vilkaar,
                                                 boolean oppfylt,
                                                 Set<String> begrunnelseKoder) {
        return lagVilkaarsresultat(behandlingsresultat, vilkaar, oppfylt, begrunnelseKoder,
            null, null);
    }

    private Vilkaarsresultat lagVilkaarsresultat(Behandlingsresultat behandlingsresultat,
                                                 Vilkaar vilkaar,
                                                 boolean oppfylt,
                                                 Set<String> begrunnelseKoder,
                                                 String begrunnelseFritekst,
                                                 String begrunnelseFritekstEngelsk) {
        Vilkaarsresultat vilkaarsresultat = new Vilkaarsresultat();
        vilkaarsresultat.setBehandlingsresultat(behandlingsresultat);
        vilkaarsresultat.setVilkaar(vilkaar);
        vilkaarsresultat.setOppfylt(oppfylt);
        vilkaarsresultat.setBegrunnelser(begrunnelseKoder.stream().map(kode -> lagBegrunnelse(vilkaarsresultat, kode))
            .collect(Collectors.toSet()));
        vilkaarsresultat.setBegrunnelseFritekst(begrunnelseFritekst);
        vilkaarsresultat.setBegrunnelseFritekstEessi(begrunnelseFritekstEngelsk);
        return vilkaarsresultat;
    }

    private VilkaarBegrunnelse lagBegrunnelse(Vilkaarsresultat vilkaarsresultat, String begrunnelseKode) {
        VilkaarBegrunnelse begrunnelse = new VilkaarBegrunnelse();
        begrunnelse.setVilkaarsresultat(vilkaarsresultat);
        begrunnelse.setKode(begrunnelseKode);
        return begrunnelse;
    }
}

File: service/src/main/kotlin/no/nav/melosys/service/behandling/VilkaarsresultatService.kt
package no.nav.melosys.service.behandling

import io.getunleash.Unleash
import no.nav.melosys.domain.Behandlingsresultat
import no.nav.melosys.domain.VilkaarBegrunnelse
import no.nav.melosys.domain.Vilkaarsresultat
import no.nav.melosys.domain.kodeverk.Kodeverk
import no.nav.melosys.domain.kodeverk.Vilkaar
import no.nav.melosys.exception.FunksjonellException
import no.nav.melosys.exception.IkkeFunnetException
import no.nav.melosys.featuretoggle.ToggleName
import no.nav.melosys.repository.BehandlingsresultatRepository
import no.nav.melosys.service.saksbehandling.SaksbehandlingRegler
import no.nav.melosys.service.vilkaar.VilkaarDto
import org.springframework.stereotype.Service
import org.springframework.transaction.annotation.Transactional


@Service
class VilkaarsresultatService(
    private val behandlingsresultatRepository: BehandlingsresultatRepository,
    private val saksbehandlingRegler: SaksbehandlingRegler,
    private val unleash: Unleash
) {
    private fun hentBehandlingsresultat(behandlingsid: Long): Behandlingsresultat =
        behandlingsresultatRepository.findById(behandlingsid)
            .orElseThrow { IkkeFunnetException("Kan ikke finne behandlingsresultat for behandling: $behandlingsid") }

    @Transactional(readOnly = true)
    fun hentVilkaar(behandlingID: Long): List<VilkaarDto> =
        hentBehandlingsresultat(behandlingID).vilkaarsresultater.map {
            VilkaarDto().apply {
                vilkaar = it.vilkaar.kode
                isOppfylt = it.isOppfylt
                begrunnelseKoder = it.begrunnelser.map { it.kode }.toSet()
                begrunnelseFritekst = it.begrunnelseFritekst
                begrunnelseFritekstEngelsk = it.begrunnelseFritekstEessi
            }
        }

    @Transactional(readOnly = true)
    fun finnVilkaarsresultat(behandlingID: Long, vilkaar: Vilkaar): Vilkaarsresultat? =
        hentBehandlingsresultat(behandlingID).vilkaarsresultater.firstOrNull { it.vilkaar == vilkaar }

    @Transactional(readOnly = true)
    fun finnUtsendingArbeidstakerVilkaarsresultat(behandlingID: Long): Vilkaarsresultat? {
        val utsendingsvilkår =
            if (unleash.isEnabled(ToggleName.MELOSYS_KONVENSJON_EFTA_LAND_OG_STORBRITANNIA)) listOf(
                Vilkaar.FO_883_2004_ART12_1,
                Vilkaar.KONV_EFTA_STORBRITANNIA_ART14_1,
                Vilkaar.KONV_EFTA_STORBRITANNIA_ART16_1,
            ) else listOf(
                Vilkaar.FO_883_2004_ART12_1,
            )

        return hentBehandlingsresultat(behandlingID).vilkaarsresultater.firstOrNull { it.vilkaar in utsendingsvilkår }
    }

    @Transactional(readOnly = true)
    fun finnUtsendingNæringsdrivendeVilkaarsresultat(behandlingID: Long): Vilkaarsresultat? {
        val utsendingsvilkår =
            if (unleash.isEnabled(ToggleName.MELOSYS_KONVENSJON_EFTA_LAND_OG_STORBRITANNIA)) listOf(
                Vilkaar.FO_883_2004_ART12_2,
                Vilkaar.KONV_EFTA_STORBRITANNIA_ART14_2,
                Vilkaar.KONV_EFTA_STORBRITANNIA_ART16_3,
            ) else listOf(
                Vilkaar.FO_883_2004_ART12_2
            )

        return hentBehandlingsresultat(behandlingID).vilkaarsresultater.firstOrNull { it.vilkaar in utsendingsvilkår }
    }

    @Transactional(readOnly = true)
    fun finnUnntaksVilkaarsresultat(behandlingID: Long): Vilkaarsresultat? {
        val unntaksvilkår =
            if (unleash.isEnabled(ToggleName.MELOSYS_KONVENSJON_EFTA_LAND_OG_STORBRITANNIA)) listOf(
                Vilkaar.FO_883_2004_ART16_1,
                Vilkaar.KONV_EFTA_STORBRITANNIA_ART18_1
            ) else listOf(
                Vilkaar.FO_883_2004_ART16_1
            )

        return hentBehandlingsresultat(behandlingID).vilkaarsresultater.firstOrNull { it.vilkaar in unntaksvilkår }
    }

    @Transactional(readOnly = true)
    fun oppfyllerVilkaar(behandlingID: Long, vilkaar: Vilkaar): Boolean =
        hentBehandlingsresultat(behandlingID).vilkaarsresultater.any { vilkaar == it.vilkaar && it.isOppfylt }

    @Transactional(readOnly = true)
    fun oppfyllerVilkaar(behandlingID: Long, vilkaar: List<Vilkaar>): Boolean =
        hentBehandlingsresultat(behandlingID).vilkaarsresultater.any { vilkaar.contains(it.vilkaar) && it.isOppfylt }

    @Transactional(readOnly = true)
    fun harVilkaarForUtsending(behandlingID: Long): Boolean =
        finnUtsendingArbeidstakerVilkaarsresultat(behandlingID) != null || finnUtsendingNæringsdrivendeVilkaarsresultat(behandlingID) != null

    @Transactional(readOnly = true)
    fun harVilkaarForUnntak(behandlingID: Long): Boolean = finnUnntaksVilkaarsresultat(behandlingID) != null

    @Transactional
    fun registrerVilkår(behandlingID: Long, vilkaarDtoer: List<VilkaarDto>) {
        validerVilkår(vilkaarDtoer)
        val behandlingsresultat = hentBehandlingsresultat(behandlingID)
        tilbakestillVilkårsresultatFraBehandlingsresultat(behandlingsresultat)
        // Flush fordi vi potensielt legger til samme vilkåret igjen. INSERT kommer før DELETE i Hibernate, som skaper UNIQUE constraint problemer uten flush.
        behandlingsresultatRepository.saveAndFlush(behandlingsresultat)
        for (vilkaarDto in vilkaarDtoer) {
            val vilkaarsresultat = lagVilkaarsresultat(
                behandlingsresultat,
                Vilkaar.valueOf(vilkaarDto.vilkaar),
                vilkaarDto.isOppfylt,
                vilkaarDto.begrunnelseKoder,
                vilkaarDto.begrunnelseFritekst,
                vilkaarDto.begrunnelseFritekstEngelsk
            )
            behandlingsresultat.vilkaarsresultater.add(vilkaarsresultat)
        }
        behandlingsresultatRepository.save(behandlingsresultat)
    }

    @Transactional
    fun tilbakestillVilkårsresultatFraBehandlingsresultatMedSaveAndFlush(behandlingID: Long) {
        val behandlingsresultat = hentBehandlingsresultat(behandlingID)
        tilbakestillVilkårsresultatFraBehandlingsresultat(behandlingsresultat)
        behandlingsresultatRepository.saveAndFlush(behandlingsresultat)
    }

    fun tilbakestillVilkårsresultatFraBehandlingsresultat(behandlingsresultat: Behandlingsresultat) {
        val behandling = behandlingsresultat.behandling
        if (behandling.fagsak.erSakstypeEøs() && !saksbehandlingRegler.harIngenFlyt(behandling)) {
            behandlingsresultat.vilkaarsresultater.removeIf { it.vilkaar !in IMMUTABLE_VILKAAR }
        } else {
            behandlingsresultat.vilkaarsresultater.clear()
        }
    }

    private fun validerVilkår(vilkaarDtoer: List<VilkaarDto>) {
        val nyeVilkår = vilkaarDtoer.map { it.vilkaar }
        for (immutableVilkår in IMMUTABLE_VILKAAR) {
            if (immutableVilkår.kode in nyeVilkår) {
                throw FunksjonellException("Kan ikke endre vilkår $immutableVilkår")
            }
        }
    }

    @Transactional
    fun oppdaterVilkaarsresultat(
        behandlingID: Long,
        vilkaar: Vilkaar,
        oppfylt: Boolean,
        begrunnelseKoder: Set<Kodeverk>
    ) {
        val behandlingsresultat = hentBehandlingsresultat(behandlingID)
        behandlingsresultat.vilkaarsresultater.clear()
        // Flush fordi vi potensielt legger til samme vilkåret igjen. INSERT kommer før DELETE i Hibernate, som skaper UNIQUE constraint problemer uten flush.
        behandlingsresultatRepository.saveAndFlush(behandlingsresultat)
        val vilkaarsresultat = lagVilkaarsresultat(
            behandlingsresultat,
            vilkaar,
            oppfylt,
            begrunnelseKoder.map { it.kode }.toSet()
        )
        behandlingsresultat.vilkaarsresultater.add(vilkaarsresultat)
        behandlingsresultatRepository.save(behandlingsresultat)
    }

    private fun lagVilkaarsresultat(
        behandlingsresultat: Behandlingsresultat,
        vilkaar: Vilkaar,
        oppfylt: Boolean,
        begrunnelseKoder: Set<String>,
        begrunnelseFritekst: String? = null,
        begrunnelseFritekstEngelsk: String? = null
    ): Vilkaarsresultat =
        Vilkaarsresultat().apply {
            this.behandlingsresultat = behandlingsresultat
            this.vilkaar = vilkaar
            this.isOppfylt = oppfylt
            this.begrunnelser = begrunnelseKoder.map { lagBegrunnelse(this, it) }.toSet()
            this.begrunnelseFritekst = begrunnelseFritekst
            this.begrunnelseFritekstEessi = begrunnelseFritekstEngelsk
        }

    private fun lagBegrunnelse(vilkaarsresultat: Vilkaarsresultat, begrunnelseKode: String) = VilkaarBegrunnelse().apply {
        this.vilkaarsresultat = vilkaarsresultat
        this.kode = begrunnelseKode
    }

    companion object {
        val IMMUTABLE_VILKAAR: Collection<Vilkaar> = setOf(Vilkaar.FO_883_2004_INNGANGSVILKAAR)
    }
}

